docs/cudf/source/developer_guide/library_design.md

```diff
@@ -0,0 +1,268 @@
+Since efficient GPU algorithms require preexisting knowledge of data layouts,
+cuDF does not support the arbitrary `object` dtype, but instead defines a few custom types for common use-cases:
+- `ListDtype`: Lists where each element in every list in a Column is of the same type
+- `StructDtype`: Dicts where a given key always maps to values of the same type
+- `CategoricalDtype`: Analogous to the pandas categorical dtype except that the categories are stored in device memory
+- `DecimalDtype`: Fixed-point numbers
+- `IntervalDtype`: Intervals
+
+Note that there is a many-to-one mapping between data types and `Column` classes.
+For instance, all numerical types (floats and ints of different widths) are all managed using `NumericalColumn`.
+
+
+### Buffer
+
+Although a `Column` represents an Arrow-compliant data structure, it does not directly handle memory mangaement.
+That job is delegated to the `Buffer` class, which represents a device memory allocation owned by another object.
+A `Buffer` constructed from a preexisting device memory allocation (such as a CuPy array) will view that memory.
+Conversely, when constructed from a host object,
+`Buffer` uses [`rmm.DeviceBuffer`](https://github.com/rapidsai/rmm#devicebuffers) to allocate new memory.
+The data is then copied from the host object into the newly allocated device memory.
+You can read more about device memory allocation with RMM [here](https://github.com/rapidsai/rmm).
+
+
+## The Cython layer
+
+The lowest level of cuDF is its interaction with `libcudf` via Cython.
+The Cython layer is composed of two components: C++ bindings and Cython wrappers.
+The first component consists of [`.pxd` files](https://cython.readthedocs.io/en/latest/src/tutorial/pxd_files.html),
+Cython declaration files that expose the contents of C++ header files to other Cython files.
+The second component consists of Cython wrappers for this functionality.
+These wrappers are necessary to expose this functionality to pure Python code.
+They also handle translating cuDF objects into their `libcudf` equivalents and invoking `libcudf` functions.
+
+Working with this layer of cuDF requires some familiarity with `libcudf`'s APIs.
+`libcudf` is built around two principal objects whose names are largely self-explanatory: `column` and `table`.
+`libcudf` also defines corresponding non-owning "view" types `column_view` and `table_view`.
+`libcudf` APIs typically accept views while returning owning types.
+
+Most cuDF Cython wrappers involve converting `cudf.Column` objects into `column_view` or `table_view` objects,
+calling a `libcudf` API with these arguments, then constructing new `cudf.Column`s from the result.
+In the context of cuDF, `libcudf`'s ownership of memory is always transient.
+cuDF's Cython wrappers immediately transfer ownership from returned owning objects to an `rmm.DeviceBuffer`.
+This behavior contrasts with how `cudf.Buffer`s work when constructed from memory allocated by other Python libraries.
```

Comment:
There's a lot that the reader needs to really grasp what we're trying to say here. I feel we should omit this for now, and really dedicate a whole other section to memory management/ownership.

Reply:
That's a fair point. I think that is out of scope for this PR, though, so would you rather I leave this for now or just remove it and plan for a future section that covers this info?

Reply:
Discussed offline, removing for now.